### PR TITLE
SG-135 Fix connection handling in mesh-io-tcpserver

### PR DESCRIPTION
**Pull request by jakubwitowski-silvair**

### CODE REVIEW

File: mesh/silvair-io.c [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+	/* Max supported version: AA.BB.CC-rcDD-12345678\0*/
+	char		silvair_version[23];
+	uint32_t	reset_reason;
+	uint32_t	uptime;
+	uint8_t		last_fault_len;
+	uint8_t		last_fault[128];
+} __packed;
+
+
+static bool io_write(struct silvair_io *io,
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c]: No need for forward declaration

File: mesh/silvair-io.c [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+	case SILVAIR_EVT_RX:
+		process_evt_rx(io, pkt_hdr, len, user_data);
+		break;
+
+	case SILVAIR_CMD_KEEP_ALIVE:
+		l_info("KEEP ALIVE RECEIVED");
+		if(disconnect_tmr) {
+			l_timeout_remove(disconnect_tmr);
+			disconnect_tmr = NULL;
+			process_evt_keep_alive(io, pkt_hdr, len);
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c]: This should be outside of if

File: mesh/silvair-io.c [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+	case SILVAIR_CMD_TX:
+	case SILVAIR_CMD_RX:
+	case SILVAIR_CMD_BOOTLOADER:
+	case SILVAIR_CMD_FILTER:
+		break;
+	}
+}
+
+static void process_slip(struct silvair_io *io,
+				struct slip *slip,
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c]: No need to pass this

File: mesh/silvair-io.c [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+		return;
+	}
+}
+
+struct silvair_io *silvair_io_new(int fd,
+				keep_alive_tmout_cb tmout_cb,
+				bool kernel_support,
+				process_packet_cb rx_cb,
+				void *context,
+				l_io_destroy_cb_t io_read_failed_cb,
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c]: Why do we need this? Isn't io_disconnect_cb enough?

File: mesh/silvair-io.c [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+	if (io->l_io)
+		l_io_destroy(io->l_io);
+
+	if (io->keep_alive_watchdog)
+		l_timeout_remove(io->keep_alive_watchdog);
+
+	if (disconnect_tmr)
+		l_timeout_remove(disconnect_tmr);
+
+	io->context = NULL;
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.c]: Not needed.

File: mesh/mesh-io-uart.c [hunk 6]
```diff
@@ -0,0 +1,611 @@
...
+static bool uart_io_destroy(struct mesh_io *mesh_io)
+{
+	struct mesh_io_private *pvt = mesh_io->pvt;
+	struct silvair_io *silvair_io = mesh_io->pvt->silvair_io;
+
+	if (!pvt)
+		return true;
+
+	if (silvair_io) {
+		l_io_destroy(silvair_io->l_io);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-uart.c]: Use silvair_io_destroy

File: mesh/mesh-io-uart.c [hunk 7]
```diff
@@ -0,0 +1,611 @@
...
+		return;
+
+	send_flush(mesh_io);
+}
+
+static void keep_alive_error(struct l_timeout *timeout, void *user_data)
+{
+	(void)timeout;
+	(void)user_data;
+	l_error("USB cable disconnected !");
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-uart.c]: Kill the whole mesh_io here

File: mesh/mesh-io-tcpserver.c [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+	close(fd);
+}
+
+static void io_disconnect_callback(void *user_data)
+{
+	struct silvair_io *silvair_io = user_data;
+	struct mesh_io *mesh_io = silvair_io->context;
+
+	if (!l_queue_remove(mesh_io->pvt->client_io, silvair_io)) {
+		perror("l_queue_remove() error");
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c:196]: This should never happen. `abort()`?

File: mesh/mesh-io-tcpserver.c [hunk 9]
```diff
@@ -0,0 +1,637 @@
...
+	struct sockaddr_in clientaddr;
+	unsigned int client_addrlen;
+
+	/* Newly accepted socket descriptor */
+	int newfd;
+
+	/* Listening socket descriptor */
+	int server_fd;
+
+	server_fd = l_io_get_fd(mesh_io->pvt->server_io);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: `l_io_get_fd(l_io)`

File: mesh/mesh-io-tcpserver.c [hunk 10]
```diff
@@ -0,0 +1,637 @@
...
+		l_error("silvair_io_new error");
+		return false;
+	}
+
+	l_io_set_close_on_destroy(silvair_io->l_io, true);
+
+	l_queue_push_tail(mesh_io->pvt->client_io, silvair_io);
+
+	/* Send keep alive request */
+	silvair_process_tx(silvair_io, NULL, 0, PACKET_TYPE_KEEP_ALIVE);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: I think this can be scheduled in `silvair_io` constructor.

File: mesh/mesh-io-tcpserver.c [hunk 11]
```diff
@@ -0,0 +1,637 @@
...
+{
+	struct silvair_io *silvair_io = user_data;
+	int fd;
+
+	(void)timeout;
+
+	if (!silvair_io)
+		return;
+
+	fd = l_io_get_fd(silvair_io->l_io);
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: Add `silvair_io_get_fd` to avoid accessing `silvair_io` fields directly.

File: mesh/mesh-io-tcpserver.c [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+	struct silvair_io *silvair_io = user_data;
+	int fd = l_io_get_fd(silvair_io->l_io);
+
+	get_fd_info(fd, "Disconnecting the TCP client", IO_TYPE_CLIENT);
+
+	/* close will trigger the io_disconnect_callback() */
+	close(fd);
+}
+
+static void io_disconnect_callback(void *user_data)
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io-tcpserver.c]: I would move this function inside `silvair-io`, extract `->context` there and then call callback provided by `mesh_io` layer in `silvair_io` constructor.

The idea is to make `silvaie_io` structure opaque from `mesh_io` point of view.

File: mesh/mesh-io.h [hunk 13]
```diff
@@ -28,7 +28,9 @@ struct mesh_io;
 
 enum mesh_io_type {
 	MESH_IO_TYPE_NONE = 0,
-	MESH_IO_TYPE_GENERIC
+	MESH_IO_TYPE_GENERIC,
+	MESH_IO_TYPE_SILVAIR,
```
Comment (michallowasrzechonek-silvair) [inline mesh/mesh-io.h]: `MESH_IO_TYPE_UART`

File: mesh/silvair-io.h [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+				process_packet_cb rx_cb,
+				void *context,
+				l_io_destroy_cb_t io_read_failed_cb,
+				l_io_destroy_cb_t io_disconnect_cb);
+
+void silvair_io_destroy(struct silvair_io *io);
+
+void silvair_io_keep_alive_wdt_refresh(struct silvair_io *io);
+
+void silvair_process_tx(struct silvair_io *io,
```
Comment (michallowasrzechonek-silvair) [inline mesh/silvair-io.h]: All public symbols should have a common `silvair_io_` prefix.

File: mesh/silvair-io.c (hunk 4 above)
Comment (jakubwitowski-silvair) [reply mesh/silvair-io.c]: in io_read_callback_destroy(), the shutdown() is called. This provides that 'generic' silvair-io.c doesn't have to know about the LL comunication protocol.

**Pull request merged**